### PR TITLE
cmd/snap-preseed: add snap-preseed executable

### PR DESCRIPTION
**Pull request by stolowski**
Introduce snap-preseed helper command line tool. The tool needs to be run against a target chroot (e.g. a mounted ubuntu-cloud image) . It's job is to find the target snapd (i.e. one coming from seeds/ inside chroot), mount it and and execute snapd in the new "pre-seeding" mode (controlled by SNAPD_PRESEED env var passed to snapd). The actual logic related to pre-seeding is implemented separately (followup PRs).

### CODE REVIEW

File: cmd/snap-start-preseed/export_test.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
Comment (pedronis) [inline cmd/snap-start-preseed/export_test.go]: let's just call it  snap-preseed, we invoke snapd and stay around anyway, so it's not just starting...

File: cmd/snap-start-preseed/main.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/jessevdk/go-flags"
+)
+
+const (
+	shortHelp = "Pre-seed snaps in an Ubuntu chroot"
```
Comment (pedronis) [inline cmd/snap-start-preseed/main.go]: Prerun the first boot seeding of snaps in a chroot with a snapd seed.

File: cmd/snap-start-preseed/main.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	longHelp  = `
+	Pre-seed snaps in an Ubuntu-based chroot directory which is expected to be a mounted image, ready to seed (/var/lib/snapd/seed directory set up for seeding)
```
Comment (pedronis) [inline cmd/snap-start-preseed/main.go]: similarly

File: cmd/snap-start-preseed/main.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	if err := checkChroot(chrootDir); err != nil {
+		return err
+	}
+
+	cleanup, err := prepareChroot(chrootDir)
+	if err != nil {
+		return err
+	}
+
+	err = startPrebakeMode(chrootDir)
```
Comment (pedronis) [inline cmd/snap-start-preseed/main.go]: let's s/prebake/preseed/ everywhere

File: cmd/snap-start-preseed/preseed_linux.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/timings"
+)
+
+var (
+	// mountPath is where target core/snapd is going to be mounted in the target chroot
+	mountPath     = "/snapd-prebake"
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: /snapd-preseed

File: cmd/snap-start-preseed/preseed_linux.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
+		return "", err
+	}
+
+	var coreSnapPath string
+	ess := seed.EssentialSnaps()
+	// TODO: handle core18, snapd snap.
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: we can use seed.UsesSnapdSnap to know which case we are in. We should probably error for now if it's true.

File: cmd/snap-start-preseed/preseed_linux.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	return coreSnapPath, nil
+}
+
+func prepareChroot(prebakeChroot string) (func(), error) {
+	if err := syscallChroot(prebakeChroot); err != nil {
+		return nil, fmt.Errorf("cannot chroot into %s: %v", prebakeChroot, err)
+	}
+
+	// GlobalRootDir is now relative to chroot env
+	rootDir := dirs.GlobalRootDir
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: I would prefer not to use dirs.GlobarRootDir here. We are not using dirs much here anyway.

File: cmd/snap-start-preseed/preseed_linux.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		path := filepath.Join(prebakeChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var systemSnapFromSeeds = func(rootDir string) (string, error) {
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: *FromSeed

File: cmd/snap-start-preseed/preseed_linux.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+				return
+			}
+			time.Sleep(time.Second)
+		}
+	}
+
+	cmd := exec.Command("mount", "-t", "squashfs", coreSnapPath, where)
+	if err := cmd.Run(); err != nil {
+		removeMountpoint()
+		return nil, fmt.Errorf("cannot mount %s at %s in pre-bake mode: %v ", coreSnapPath, where, err)
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: s/prebake/preseed/

File: cmd/snap-start-preseed/preseed_linux.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+	return func() {
+		unmount()
+		removeMountpoint()
+	}, nil
+}
+
+// startPrebakeMode runs snapd in a prebake mode. It assumes running in a chroot.
+// The chroot is expected to be set-up and ready to use (critical system directories mounted).
+func startPrebakeMode(prebakeChroot string) error {
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: s/startPrebakeMode/runPreseedMode/ ?

File: cmd/snap-start-preseed/preseed_linux.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	// exec snapd relative to new chroot, e.g. /snapd-prebake/usr/lib/snapd/snapd
+	path := filepath.Join(mountPath, "/usr/lib/snapd/snapd")
+
+	// run snapd in pre-baking mode
+	cmd := exec.Command(path)
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "SNAPD_PREBAKE_IMAGE=1")
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: just SNAPD_PRESEED  , we might switch to command line argument later maybe, but env var it's ok for now

File: cmd/snap-start-preseed/preseed_linux.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+	cmd.Stderr = Stderr
+	cmd.Stdout = Stdout
+
+	fmt.Printf("starting pre-baking mode: %s\n", prebakeChroot)
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: preseed

File: cmd/snap-start-preseed/preseed_linux.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("image-prebaking error: %v\n", err)
```
Comment (pedronis) [inline cmd/snap-start-preseed/preseed_linux.go]: error running snapd in preseed mode:

File: packaging/ubuntu-16.04/snapd.install [hunk 14]
```diff
@@ -8,6 +8,7 @@ usr/bin/snap-update-ns /usr/lib/snapd/
 usr/bin/snapd /usr/lib/snapd/
 usr/bin/snap-seccomp /usr/lib/snapd/
 usr/bin/snap-recovery /usr/lib/snapd/
+usr/bin/snap-preseed /usr/lib/snapd/
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.install]: Is this required to be present in the classic package?

How is this going to be used?

**Comment by pedronis** [comment]
It is fine if this new executable is packaged only for Ubuntu (at least initially).

File: packaging/ubuntu-16.04/snapd.install (hunk 14 above)
Comment (stolowski) [reply packaging/ubuntu-16.04/snapd.install]: It is going to be used on classic ubuntu systems only when preparing ubuntu cloud images (at least for the foreseeable future), it will be run against a mounted image. So it's a very specialized use case. Incorporating it into snap command is a possiblity, but for the time being the command is going to be usable for a very specific usecase only.

File: cmd/snap-preseed/main.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+	chrootDir := rest[0]
+	if err := checkChroot(chrootDir); err != nil {
+		return err
+	}
+
+	cleanup, err := prepareChroot(chrootDir)
+	if err != nil {
+		return err
+	}
+
```
Comment (bboozzoo) [inline cmd/snap-preseed/main.go]:
```
// executing inside the chroot
```

File: cmd/snap-preseed/preseed_linux.go [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	return coreSnapPath, nil
+}
+
+func prepareChroot(preseedChroot string) (func(), error) {
+	if err := syscallChroot(preseedChroot); err != nil {
+		return nil, fmt.Errorf("cannot chroot into %s: %v", preseedChroot, err)
+	}
+
```
Comment (bboozzoo) [inline cmd/snap-preseed/preseed_linux.go]: IIRC chroot does not change pwd, so we should do:
```
if err := os.Chdir("/"); err != nil {
    return fmt.Errorf("cannot chdir to /: %v", err)
}
```

File: cmd/snap-preseed/preseed_linux.go [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+	// GlobalRootDir is now relative to chroot env
+	rootDir := dirs.GlobalRootDir
```
Comment (bboozzoo) [inline cmd/snap-preseed/preseed_linux.go]: After os.Chdir(), we should be able to do `dirs.SetRootDir("/")`, from this point on, dirs.SnapSeedDir should be set correctly.

File: cmd/snap-preseed/preseed_linux.go [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+	unmount := func() {
+		fmt.Fprintf(Stdout, "unmounting: %s\n", mountPath)
+		cmd := exec.Command("umount", mountPath)
+		if err := cmd.Run(); err != nil {
+			fmt.Fprintf(Stderr, "%v", err)
+		}
+	}
+
+	return func() {
+		unmount()
```
Comment (bboozzoo) [inline cmd/snap-preseed/preseed_linux.go]: Should this exit the chroot and reset dirs.RootDir?

For exiting the chroot, you'll probably need to grab a file descriptor to `"/"` before chrooting and `f.Chdir()` to it in the cleanup.

File: cmd/snap-preseed/preseed_linux.go (hunk 17 above)
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: Done. FWTW it was following `image.go` with `dirs.SnapSeedDirUnder(rootDir)`; setting root is not critical since we are only processing seed directory, but I've no strong opinion.

File: cmd/snap-preseed/preseed_linux.go (hunk 18 above)
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: I don't think we should do that. This is a standalone utility and doing chroot work is its last operation. The cleanup functions are meant to clean up any artifacts from the filesystem.

File: cmd/snap-preseed/main.go [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+	"io"
+	"os"
+
+	"github.com/jessevdk/go-flags"
+)
+
+const (
+	shortHelp = "Prerun the first boot seeding of snaps in a chroot with a snapd seed."
+	longHelp  = `
+	Prerun the first boot seeding of snaps in an Ubuntu-based chroot directory with a snapd seed (/var/lib/snapd/seed directory set up for seeding).
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap-preseed/main.go]: Done.

File: cmd/snap-preseed/main.go (hunk 15 above)
Comment (stolowski) [reply cmd/snap-preseed/main.go]: Done.

File: cmd/snap-preseed/preseed_linux.go [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+var systemSnapFromSeeds = func(rootDir string) (string, error) {
+	seedDir := filepath.Join(dirs.SnapSeedDirUnder(rootDir))
+	seed, err := seed.Open(seedDir, "")
+	if err != nil {
+		return "", err
+	}
+
+	if err := seed.LoadAssertions(nil, nil); err != nil {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: Done.

File: cmd/snap-preseed/preseed_linux.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+		return "", err
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: I just realized state doesn't exist at this point until snapd is started for the first time (in presseding mode). I'd therefore leave it for a followup as we need to ensure that an almost-empty state is fine when snapd is started for preseeding.

File: cmd/snap-preseed/preseed_linux.go [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var systemSnapFromSeeds = func() (string, error) {
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go]: FromSeed

File: cmd/snap-preseed/preseed_linux.go [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
+		return "", err
+	}
+
+	var coreSnapPath string
+	ess := seed.EssentialSnaps()
+	// TODO: handle core18, snapd snap.
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go]: my comment about checking and erroring based on seed.UsesSnapdSnap seems to have got lost

File: cmd/snap-preseed/preseed_linux.go [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+	for _, snap := range ess {
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go]: we can assume that core/snapd is ess[0]

File: cmd/snap-preseed/preseed_linux.go (hunk 17 above)
Comment (pedronis) [reply cmd/snap-preseed/preseed_linux.go]: I find the SetRootDir a bit confusing personally,  we don't set it before to something else? the point of a  chroot is indeed to change / at which both dirs package values are correct

Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: I've reverted this change.

**Comment by bboozzoo** [comment]
It'd be nice to add a spread test in a follow up.

File: cmd/snap-preseed/main.go [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+		return fmt.Errorf("must be run as root")
+	}
+
+	rest, err := parser.ParseArgs(args)
+	if err != nil {
+		return err
+	}
+
+	if len(rest) == 0 {
+		return fmt.Errorf("need the chroot path as argument")
```
Comment (pedronis) [inline cmd/snap-preseed/main.go]: s/the//

File: cmd/snap-preseed/preseed_linux.go [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/timings"
+)
+
+var (
+	// mountPath is where target core/snapd is going to be mounted in the target chroot
+	mountPath     = "/snapd-preseed"
```
[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap-preseed/preseed_linux.go]: that might be cleaner indeed

[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap-preseed/preseed_linux.go]: Also

s/mountPath/snapdMountPath/

for clarity

File: cmd/snap-preseed/preseed_linux.go [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var systemSnapFromSeed = func(rootDir string) (string, error) {
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go]: this implementation is not tested afaict

File: packaging/ubuntu-14.04/snapd.install [hunk 28]
```diff
@@ -6,6 +6,7 @@ usr/bin/snap-repair /usr/lib/snapd/
 usr/bin/snap-failure /usr/lib/snapd/
 usr/bin/snap-update-ns /usr/lib/snapd/
 usr/bin/snapd /usr/lib/snapd/
+usr/bin/snap-preseed /usr/lib/snapd/
```
Comment (pedronis) [inline packaging/ubuntu-14.04/snapd.install]: wondering if we should wait to do this until we can have a spread test and is (almost) fully working?

Comment (stolowski) [reply packaging/ubuntu-14.04/snapd.install]: No problem, makes sense, done.

File: cmd/snap-preseed/preseed_linux.go (hunk 27 above)
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: Right, added unit tests.

**Comment by stolowski** [comment]
> It'd be nice to add a spread test in a follow up.

Absolutely, I'll start working on it soon.

File: cmd/snap-preseed/main.go [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+	"io"
+	"os"
+
+	"github.com/jessevdk/go-flags"
+)
+
+const (
+	shortHelp = "Prerun the first boot seeding of snaps in a chroot with a snapd seed."
+	longHelp  = `
+	Prerun the first boot seeding of snaps in an Ubuntu-based chroot directory
```
Comment (zyga) [inline cmd/snap-preseed/main.go]: I think this needs a rewrite. It reads as a techno-jargon that is hard to follow.
```
The prerun command improves performance of first boot seeding process by
pre-computing and caching the result of certain operations.
```

File: cmd/snap-preseed/preseed_linux.go [hunk 30]
```diff
@@ -0,0 +1,188 @@
...
+// be exhaustive check, but one that prevents running the tool against a wrong directory
+// by an accident, which would lead to hard to understand errors from snapd in preseed
+// mode.
+func checkChroot(preseedChroot string) error {
+	exists, isDir, err := osutil.DirExists(preseedChroot)
+	if err != nil {
+		return fmt.Errorf("cannot verify target chroot directory %s: %v", preseedChroot, err)
+	}
+	if !exists || !isDir {
+		return fmt.Errorf("target chroot directory %s doesn't exist or is not a directory", preseedChroot)
```
Comment (zyga) [inline cmd/snap-preseed/preseed_linux.go]:
```suggestion
		return fmt.Errorf("cannot verify %q: is not a directory", preseedChroot)
```

File: cmd/snap-preseed/preseed_linux.go [hunk 31]
```diff
@@ -0,0 +1,188 @@
...
+
+// checkChroot does a basic sanity check of the target chroot environment, e.g. makes
+// sure critical virtual filesystems (such as proc) are mounted. This is not meant to
+// be exhaustive check, but one that prevents running the tool against a wrong directory
+// by an accident, which would lead to hard to understand errors from snapd in preseed
+// mode.
+func checkChroot(preseedChroot string) error {
+	exists, isDir, err := osutil.DirExists(preseedChroot)
+	if err != nil {
+		return fmt.Errorf("cannot verify target chroot directory %s: %v", preseedChroot, err)
```
Comment (zyga) [inline cmd/snap-preseed/preseed_linux.go]:
```suggestion
		return fmt.Errorf("cannot verify %q: %v", preseedChroot, err)
```

File: cmd/snap-preseed/preseed_linux.go [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	if !exists || !isDir {
+		return fmt.Errorf("target chroot directory %s doesn't exist or is not a directory", preseedChroot)
+	}
+
+	// sanity checks of the critical mountpoints inside chroot directory
+	for _, p := range []string{"/sys/kernel/security/apparmor", "/proc/self", "/dev/mem"} {
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
```
Comment (zyga) [inline cmd/snap-preseed/preseed_linux.go]:
```suggestion
			return fmt.Errorf("cannot pre-seed without access to %q", path)
```

File: cmd/snap-preseed/preseed_linux.go [hunk 33]
```diff
@@ -0,0 +1,188 @@
...
+		return nil, err
+	}
+
+	// create mountpoint for core/snapd
+	where := filepath.Join(rootDir, snapdMountPath)
+	if err := os.MkdirAll(where, 0755); err != nil {
+		return nil, err
+	}
+
+	removeMountpoint := func() {
```
Comment (zyga) [inline cmd/snap-preseed/preseed_linux.go]: Why is retry-and-wait approach required? Can you add that as a code comment please.

File: cmd/snap-preseed/preseed_other.go [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+)
+
```
Comment (zyga) [inline cmd/snap-preseed/preseed_other.go]: This is nice :)

File: packaging/debian-sid/rules [hunk 35]
```diff
@@ -252,6 +252,9 @@ override_dh_install-arch:
 	rm $(CURDIR)/debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/snapd.system-shutdown.service
 	rm $(CURDIR)/debian/snapd/usr/lib/snapd/snapd.run-from-snap
 
+	# snap-preseed is only useful on ubuntu
```
Comment (zyga) [inline packaging/debian-sid/rules]: Why is is only useful on Ubuntu? Can you add a code comment please.

File: packaging/ubuntu-16.04/rules [hunk 36]
```diff
@@ -253,6 +253,9 @@ override_dh_install:
 	rm $(CURDIR)/debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/snapd.apparmor.service
 	rm $(CURDIR)/debian/tmp/usr/lib/snapd/snapd-apparmor
 
+	# snap-preseed should not be distributed yet.
```
Comment (zyga) [inline packaging/ubuntu-16.04/rules]: Can you please also change `packaging/snapd.mk`

File: cmd/snap-preseed/main.go (hunk 29 above)
Comment (pedronis) [reply cmd/snap-preseed/main.go]: The audience of this tool is not end users.

Comment (pedronis) [reply cmd/snap-preseed/main.go]: We can try to make the sentences less clunky but no point hiding details.

Comment (pedronis) [reply cmd/snap-preseed/main.go]: Maybe @degville can help a bit.

Comment (pedronis) [reply cmd/snap-preseed/main.go]: Maybe 

```
The snap-preseed command takes a directory with an image filesytem including seed snaps (at /var/lib/snapd/seed) and pre-runs as much as possible of first boot snap installation to optimize it.
```

Comment (zyga) [reply cmd/snap-preseed/main.go]: @pedronis this is very nice, let's use it.

File: cmd/snap-preseed/preseed_linux.go (hunk 33 above)
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go]: I've dropped it. As mentioned on the standup it was most likely a leftover from early version of this tool where some other problem (related to unmounting) prevented remove.

File: packaging/ubuntu-16.04/rules (hunk 36 above)
Comment (stolowski) [reply packaging/ubuntu-16.04/rules]: Done.

File: cmd/snap-preseed/main.go [hunk 37]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/jessevdk/go-flags"
+)
+
+const (
+	shortHelp = "Prerun the first boot seeding of snaps in a chroot with a snapd seed."
+	longHelp  = `
+The snap-preseed command takes a directory with an image filesytem including
+seed snaps (at /var/lib/snapd/seed) and pre-runs as much as possible without
+booting of first boot snap installation to optimize it.`
+)
```
Comment (degville) [inline cmd/snap-preseed/main.go]: I'd tweak @pedronis words slightly (especially `without booting of first boot snap installation to optimize it`, whose meaning isn't clear to me. Sorry to make another suggestion, but how about:

`The snap-preseed command takes a directory with an image filesytem, including
seed snaps (at /var/lib/snapd/seed), and pre-runs as much as possible to optimise performance without booting the boot snap installation process.`

File: cmd/snap-preseed/main.go (hunk 29 above)
Comment (degville) [reply cmd/snap-preseed/main.go]: I'm made a slight tweak suggestion to this below.

File: cmd/snap-preseed/main.go (hunk 37 above)
Comment (pedronis) [reply cmd/snap-preseed/main.go]: Sadly that rewrite is confusing in different ways,  it sounds we don't start the snap installation process, which is actually what we do. OTOH we don't do any other kind of booting. I need to think a bit.

Comment (degville) [reply cmd/snap-preseed/main.go]: Ok, it's that last part I'm having difficulty with, maybe more like:

`To improves a snap's first-boot caching performance, the snap-preseed command takes a directory with an image filesytem, including seed snaps (at /var/lib/snapd/seed), and pre-runs as much as possible, including much of the boot snap installation process itself.`

File: cmd/snap-preseed/preseed_linux.go [hunk 38]
```diff
@@ -0,0 +1,181 @@
...
+	path := filepath.Join(snapdMountPath, dirs.CoreLibExecDir, "snapd")
+
+	// run snapd in preseed mode
+	cmd := exec.Command(path)
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "SNAPD_PRESEED=1")
+	cmd.Stderr = Stderr
+	cmd.Stdout = Stdout
+
+	fmt.Fprintf(Stdout, "starting preseed: %s\n", preseedChroot)
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go]: "starting to preseed root: ..."
